rust/cyclotron-core/src/worker.rs

```diff
@@ -0,0 +1,245 @@
+use std::{collections::HashMap, sync::Arc};
+
+use chrono::{DateTime, Utc};
+use sqlx::PgPool;
+use std::sync::Mutex;
+use uuid::Uuid;
+
+use crate::{
+    base_ops::{
+        dequeue_jobs, dequeue_with_vm_state, flush_job, set_heartbeat, Job, JobState, JobUpdate,
+        WaitingOn,
+    },
+    error::QueueError,
+    PoolConfig,
+};
+
+// The worker's interface to the underlying queue system - a worker can do everything except
+// create jobs (because job creation has to be shard-aware).
+//
+// This interface looks stange, because a lot of things that would normally be done with lifetimes
+// and references are done with uuid's instead (and we lose some nice raii stuff as a result), but
+// the reason for this is that this is designed to be embedded in other runtimes, where handing out
+// lifetime'd references or things with drop impls isn't really practical. This makes it a little
+// awkward to use, but since it's meant to be the core of other abstractions, I think it's ok for
+// now (client libraries should wrap this to provide better interfaces).
+pub struct Worker {
+    pool: PgPool,
+    // All dequeued job IDs that haven't been flushed yet. The idea is this lets us
+    // manage, on the rust side of any API boundary, the "pending" update of any given
+    // job, such that a user can progressively build up a full update, and then flush it,
+    // rather than having to track the update state on their side and submit it all at once
```

Comment:
Most wont, but e.g. in the fetch impl, I have a pattern of queuing up the set of updates that would put a job into the deal letter queue, then doing some serde or other "this should never fail, if it does that's a coding error in the fetch worker" work, and if any of that causes an error I just flush the update (sending the job to the DLQ), otherwise I then queue up the updates to send it back to an available state or whatever, and flush

It looks like (ripped from in-progress code, but you maybe get the idea)
```rust
// Complete the job, either because we got a good response, or because the jobs retries
// have been exceeded.
pub async fn complete_job(
    worker: &Worker,
    job: &Job,
    return_worker: WaitingOn,
    return_queue: Option<String>,
    on_finish: OnFinish,
    result: FetchResult,
) -> Result<(), WorkerError> {
    // If we fail any serde, we just want to flush to the DLQ and bail
    worker.set_state(job.id, JobState::Available)?;
    worker.set_queue(job.id, DEAD_LETTER_QUEUE)?;

    let is_completed = result.is_completed();

    let result = match serde_json::to_string(&result) {
        Ok(r) => r,
        Err(e) => {
            // Leave behind a hint for debugging
            worker.set_metadata(job.id, Some(format!("Failed to serialise result: {}", e)))?;
            worker.flush_job(job.id).await?;
            return Err(WorkerError::SerdeError(e));
        }
    };

    worker.set_queue(
        job.id,
        &return_queue.unwrap_or_else(|| job.queue_name.clone()),
    )?;
    worker.set_waiting_on(job.id, return_worker)?;

    match (is_completed, on_finish) {
        (true, _) | (false, OnFinish::Return) => {
            worker.set_state(job.id, JobState::Available)?;
        }
        (false, OnFinish::Complete) => {
            worker.set_state(job.id, JobState::Failed)?;
        }
    }

    worker.set_parameters(job.id, Some(result))?;
    worker.set_metadata(job.id, None)?; // We're finished with the job, so clear our internal state
    worker.flush_job(job.id).await?;

    Ok(())
}
```

Reply:
Just my opinion, but following what state everything in here seems like more of a burden than just having a happy path that sets things at the end, otherwise throw an error, and have a wrapper function that catches errors and sets the state properly in that case.